source/postprocess/mass_error_statistics.cc

```diff
@@ -0,0 +1,225 @@
+
+            in.position = fe_values_support.get_quadrature_points();
+
+            // since we are not reading the viscosity and the viscosity
+            // is the only coefficient that depends on the strain rate,
+            // we need not compute the strain rate. set the corresponding
+            // array to empty, to prevent accidental use and skip the
+            // evaluation of the strain rate in evaluate().
+            in.strain_rate.resize(0);
+
+            for (unsigned int i=0; i<support_points.size(); ++i)
+              {
+                for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+                  in.composition[i][c] = composition_values[c][i];
+              }
+            in.cell = &cell;
+
+            this->get_material_model().evaluate(in, out);
+
+            std::vector<Tensor<1,dim> > velocity(n_q_points);
+            std::vector<double>         velocity_divergence(n_q_points);
+            fe_values[this->introspection().extractors.velocities].get_function_values(this->get_solution(),velocity);
+            fe_values[this->introspection().extractors.velocities].get_function_divergences(this->get_solution(),velocity_divergence);
+
+            for (unsigned int q = 0; q < n_q_points; ++q)
+              {
+                Tensor<1,dim> density_gradient = Tensor<1,dim>();
+                double        density = 0.0;
+                for (unsigned int i = 0; i < support_points.size(); ++i)
+                  {
+                    density_gradient += out.densities[i] * fe_values_density.shape_grad(i,q);
+                    density += out.densities[i] * fe_values_density.shape_value(i,q);
+                  }
+
+                const double point_error = (density * velocity_divergence[q]
+                                            + velocity[q] * density_gradient);
+
+                local_integral_mass_error += point_error * fe_values.JxW(q);
+                local_norm_mass_error += point_error * point_error * fe_values.JxW(q);
+              }
+          }
+
+      const double global_integral_mass_error
+        = Utilities::MPI::sum (local_integral_mass_error, this->get_mpi_communicator());
+      const double global_norm_mass_error
+        = std::sqrt(Utilities::MPI::sum (local_norm_mass_error, this->get_mpi_communicator()));
+
+      if (this->convert_output_to_years() == true)
+        {
+          statistics.add_value ("Integral mass error (kg/year)",
+                                global_integral_mass_error * year_in_seconds);
+          statistics.add_value ("Norm mass error (kg/year)",
+                                global_norm_mass_error * year_in_seconds);
+
+          // also make sure that the other columns filled by the this object
+          // all show up with sufficient accuracy and in scientific notation
+          {
+            const char *columns[] = { "Integral mass error (kg/year)",
+                                      "Norm mass error (kg/year)"
```

Comment:
also, if you move these two lines one block up, then you don't need to write the same text twice